spec/io/io_spec.rb

```diff
@@ -1,39 +1,70 @@
+            "http://dummy_remote_url/#{file}.csv"
+          end
+        end
       end
 
-      it "works properly for repeated headers" do
-        df = Daru::DataFrame.from_csv('spec/fixtures/repeated_fields.csv',header_converters: :symbol)
-        expect(df.vectors.to_a).to eq(["id", "name_1", "age_1", "city", "a1", "name_2", "age_2"])
+      it "loads from a CSV file" do
+        file = 'matrix_test'
+        paths = [local_csv_url(file), remote_csv_url(file)]
+        paths.each do |path|
+          df = Daru::DataFrame.from_csv(path,col_sep: ' ', headers: true)
+          df.vectors = [:image_resolution, :mls, :true_transform].to_index
+          expect(df.vectors).to eq([:image_resolution, :mls, :true_transform].to_index)
+          expect(df[:image_resolution].first).to eq(6.55779)
+          expect(df[:true_transform].first).to eq("-0.2362347,0.6308649,0.7390552,0,0.6523478,-0.4607318,0.6018043,0,0.7201635,0.6242881,-0.3027024,4262.65,0,0,0,1")
+        end
+      end
 
-        age = Daru::Vector.new([3, 4, 5, 6, nil, 8])
-        expect(df['age_2']).to eq(age)
+      it "works properly for repeated headers" do
+        file = 'repeated_fields'
+        paths = [local_csv_url(file), remote_csv_url(file)]
+        paths.each do |path|
+          df = Daru::DataFrame.from_csv(path,header_converters: :symbol)
+          expect(df.vectors.to_a).to eq(["id", "name_1", "age_1", "city", "a1", "name_2", "age_2"])
+
+          age = Daru::Vector.new([3, 4, 5, 6, nil, 8])
+          expect(df['age_2']).to eq(age)
+        end
       end
 
       it "accepts scientific notation as float" do
-        ds = Daru::DataFrame.from_csv('spec/fixtures/scientific_notation.csv', order: ['x', 'y'])
-        expect(ds.vectors.to_a).to eq(['x', 'y'])
-        y = [9.629587310436753e+127, 1.9341543147883677e+129, 3.88485279048245e+130]
-        y.zip(ds['y']).each do |y_expected, y_ds|
-          expect(y_ds).to be_within(0.001).of(y_expected)
+        file = 'scientific_notation'
+        paths = [local_csv_url(file), remote_csv_url(file)]
+        paths.each do |path|
+          ds = Daru::DataFrame.from_csv(path, order: ['x', 'y'])
+          expect(ds.vectors.to_a).to eq(['x', 'y'])
+          y = [9.629587310436753e+127, 1.9341543147883677e+129, 3.88485279048245e+130]
+          y.zip(ds['y']).each do |y_expected, y_ds|
+            expect(y_ds).to be_within(0.001).of(y_expected)
+          end
         end
       end
 
       it "follows the order of columns given in CSV" do
-        df = Daru::DataFrame.from_csv 'spec/fixtures/sales-funnel.csv'
-        expect(df.vectors.to_a).to eq(%W[Account Name Rep Manager Product Quantity Price Status])
+        file = 'sales-funnel'
+        paths = [local_csv_url(file), remote_csv_url(file)]
+        paths.each do |path|
+          df = Daru::DataFrame.from_csv path
+          expect(df.vectors.to_a).to eq(%W[Account Name Rep Manager Product Quantity Price Status])
+        end
       end
```

Comment:
>  Now, should we have 4 more tests by reading from the remote URLs

One should be enough. We know that internally it uses the same logic, just takes the data from URL, so no need to test whether it parses headers properly and so on.

> Also, wouldn't file-wise spec enhancements be easier to do, so that the specs are uniform at any given point of time?

Yes, we even have a ticket for this: https://github.com/SciRuby/daru/issues/317 without much volunteers though. I am not sure that rewriting entire 300 lines of io_spec.rb should go to this PR, as there are several concerns will be mixing. If you have a lot of enthusiasm, you can do it in the PR, though, but if not, just make your new specs pretty.